src/Evolution/DiscontinuousGalerkin/SlopeLimiters/HwenoModifiedSolution.hpp

```diff
@@ -0,0 +1,488 @@
+        get<Tag>(neighbor_and_data.second.volume_data)[tensor_index];
+
+    // Add terms from the primary neighbor
+    if (neighbor_and_data.first == primary_neighbor) {
+      for (size_t r = 0; r < neighbor_mesh.number_of_grid_points(); ++r) {
+        for (size_t s = 0; s < number_of_grid_points; ++s) {
+          b[s] += neighbor_tensor_component[r] *
+                  neighbor_quadrature_weights[r] * interpolation_matrix(r, s);
+        }
+      }
+    }
+    // Add terms from the secondary neighbors
+    else {
+      const double quadrature_weights_dot_u = [&]() noexcept {
+        double result = 0.;
+        for (size_t r = 0; r < neighbor_mesh.number_of_grid_points(); ++r) {
+          result +=
+              neighbor_tensor_component[r] * neighbor_quadrature_weights[r];
+        }
+        return result;
+      }
+      ();
+      b += quadrature_weights_dot_u *
+           quadrature_weights_dot_interpolation_matrix;
+    }
+  }
+  return b;
+}
+
+// Solve the constrained fit problem that gives the HWENO modified solution,
+// for one particular tensor component. For details, see documentation of
+// `compute_hweno_modified_neighbor_solution` below.
+template <typename Tag, size_t VolumeDim, typename Package>
+void solve_constrained_fit(
+    const gsl::not_null<DataVector*> constrained_fit_result,
+    const DataVector& u, const size_t tensor_index,
+    const Element<VolumeDim>& element, const Mesh<VolumeDim>& mesh,
+    const std::unordered_map<
+        std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>, Package,
+        boost::hash<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>>&
+        neighbor_data,
+    const std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>&
+        primary_neighbor,
+    const std::vector<std::pair<Direction<VolumeDim>, ElementId<VolumeDim>>>&
+        neighbors_to_exclude) noexcept {
+  // Get the cache of linear algebra quantities for this element
+  const ConstrainedFitCache<VolumeDim>& cache =
+      constrained_fit_cache(element, mesh);
+
+  // Because we don't support h-refinement, the direction is the only piece
+  // of the neighbor information that we actually need.
+  const Direction<VolumeDim> primary_direction = primary_neighbor.first;
+  const std::vector<Direction<VolumeDim>>
+      directions_to_exclude = [&neighbors_to_exclude]() noexcept {
+    std::vector<Direction<VolumeDim>> result(neighbors_to_exclude.size());
+    for (size_t i = 0; i < result.size(); ++i) {
+      result[i] = neighbors_to_exclude[i].first;
+    }
+    return result;
+  }
+  ();
+
+  const DataVector& w = cache.quadrature_weights;
+  const DirectionMap<VolumeDim, Matrix>& m = cache.interpolation_matrices;
+  const DirectionMap<VolumeDim, DataVector>& w_dot_m =
+      cache.quadrature_weights_dot_interpolation_matrices;
+
+  // Use cache if possible, or compute matrix if we are in edge case
+  const Matrix& a_inverse = LIKELY(directions_to_exclude.size() < 2)
```

Comment:
A question for possible future exploration (and I don't know where else to put it) : I'm curious if something can be done with the fact that the individual interpolation matrices are outer products to get away with inverting and caching things that are only ~ n^2, rather than n^6.
It won't be very direct because (if I understand correctly) A is the sum of two distinct matrices that are outer products.  Do you think this is worth pursuing? if so, we can talk math sometime.

Reply:
Jordan and I discussed this in person. Summary: I do not think it is worth pursuing at this time, because we still need to generalize the limiter to curved grids and the math may or may not evolve. We could revisit this after we find a working algorithm on curved grids.